storage: TestReplicateRemovedNodeDisruptiveElection failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/9671342fead0509bec0913bae4ae1f244660788e

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestReplicateRemovedNodeDisruptiveElection PKG=github.com/cockroachdb/cockroach/pkg/storage TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1298865&tab=buildLog

```
I190521 05:51:59.708786 802423 storage/replica_command.go:797  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3, gen=0]
I190521 05:51:59.711903 802423 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=58acab83] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
I190521 05:51:59.714853 802423 storage/store_snapshot.go:780  [s1,r1/1:/M{in-ax}] sending preemptive snapshot 81371f1d at applied index 21
I190521 05:51:59.715066 802423 storage/store_snapshot.go:823  [s1,r1/1:/M{in-ax}] streamed snapshot to (n4,s4):?: kv pairs: 60, log entries: 0, rate-limit: 8.0 MiB/sec, 0.00s
I190521 05:51:59.715430 803430 storage/replica_raftstorage.go:814  [s4,r1/?:{-}] applying preemptive snapshot at index 21 (id=81371f1d, encoded size=8699, 1 rocksdb batches, 0 log entries)
I190521 05:51:59.715702 803430 storage/replica_raftstorage.go:820  [s4,r1/?:/M{in-ax}] applied preemptive snapshot in 0ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I190521 05:51:59.716248 802423 storage/replica_command.go:797  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n4,s4):4): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=0]
I190521 05:51:59.718095 802423 storage/replica_raft.go:396  [s1,r1/1:/M{in-ax},txn=29bd4c81] proposing ADD_REPLICA((n4,s4):4): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3 (n4,s4):4] next=5
I190521 05:51:59.871116 802423 storage/replica_command.go:797  [s2,r1/2:/M{in-ax}] change replicas (REMOVE_REPLICA (n1,s1):1): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, (n3,s3):3, (n4,s4):4, next=5, gen=0]
I190521 05:51:59.874590 802423 storage/replica_raft.go:396  [s2,r1/2:/M{in-ax},txn=a986bff5] proposing REMOVE_REPLICA((n1,s1):1): updated=[(n4,s4):4 (n2,s2):2 (n3,s3):3] next=5
I190521 05:51:59.895600 804263 storage/store.go:2507  [replicaGC,s1,r1/1:/M{in-ax}] removing replica r1/1
I190521 05:51:59.896011 804263 storage/replica_destroy.go:150  [replicaGC,s1,r1/1:/M{in-ax}] removed 53 (49+4) keys in 0ms [clear=0ms commit=0ms]
I190521 05:52:01.187130 802715 gossip/gossip.go:1510  [n2] node has connected to cluster via gossip
I190521 05:52:01.454666 803192 gossip/gossip.go:1510  [n4] node has connected to cluster via gossip
I190521 05:52:02.025923 802945 gossip/gossip.go:1510  [n3] node has connected to cluster via gossip
I190521 05:52:14.904041 802581 storage/compactor/compactor.go:329  [s1,compactor] purging suggested compaction for range /Min - /Max that contains live data
I190521 05:52:44.888715 1077179 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190521 05:52:44.888740 1077179 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190521 05:52:44.888762 1077176 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190521 05:52:44.888776 1077176 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190521 05:52:44.888796 1077177 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190521 05:52:44.888808 1077177 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
I190521 05:52:44.888826 1077178 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
1      [async] intent_resolver_gc_batcher
I190521 05:52:44.888856 1077178 util/stop/stopper.go:546  quiescing; tasks left:
1      [async] intent_resolver_ir_batcher
W190521 05:52:44.890249 802715 gossip/gossip.go:1496  [n2] no incoming or outgoing connections
W190521 05:52:44.890360 803665 storage/raft_transport.go:583  while processing outgoing Raft queue to node 4: rpc error: code = Unavailable desc = transport is closing:
--- FAIL: TestReplicateRemovedNodeDisruptiveElection (45.44s)
    client_raft_test.go:3469: did not get expected ReplicaTooOldError error

```

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -3615,6 +3615,7 @@ func (s *Store) processRequestQueue(ctx context.Context, rangeID roachpb.RangeID
 	q.Unlock()
 
 	var lastRepl *Replica
+	var hadError bool
 	for i, info := range infos {
 		last := i == len(infos)-1
 		pErr := s.withReplicaForRequest(info.respStream.Context(), info.req,
@@ -3636,21 +3637,32 @@ func (s *Store) processRequestQueue(ctx context.Context, rangeID roachpb.RangeID
 				return pErr
 			})
 		if pErr != nil {
-			// If we're unable to process the request, clear the request queue. This
-			// only happens if we couldn't create the replica because the request was
-			// targeted to a removed range. This is also racy and could cause us to
-			// drop messages to the deleted range occasionally (#18355), but raft
-			// will just retry.
-			q.Lock()
-			if len(q.infos) == 0 {
-				s.replicaQueues.Delete(int64(rangeID))
-			}
-			q.Unlock()
+			hadError = true
 			if err := info.respStream.Send(newRaftMessageResponse(info.req, pErr)); err != nil {
 				// Seems excessive to log this on every occurrence as the other side
 				// might have closed.
 				log.VEventf(ctx, 1, "error sending error: %s", err)
 			}
+		}
+	}
+
+	if hadError {
+		// If we're unable to process a request, consider dropping the request queue
+		// to free up space in the map.
+		// This is relevant if requests failed because the target replica could not
+		// be created (for example due to the Raft tombstone). The particular code
+		// here takes into account that we don't want to drop the queue if there
+		// are other messages waiting on it, or if the target replica exists. Raft
+		// tolerates the occasional dropped message, but our unit tests are less
+		// forgiving.
+		//
+		// See https://github.com/cockroachdb/cockroach/issues/30951#issuecomment-428010411.
+		if _, exists := s.mu.replicas.Load(int64(rangeID)); !exists {
+			q.Lock()
+			if len(q.infos) == 0 {
+				s.replicaQueues.Delete(int64(rangeID))
+			}
+			q.Unlock()
 		}
 	}
 

